external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala

```diff
@@ -930,6 +930,33 @@ class AvroSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
     }
   }
 
+  test("support user provided non-nullable avro schema " +
```

Comment:
This doesn't quite make sense to me. Looks if the catalyst schema is nullable, it should reject non-nullable avro schema

Reply:
BTW, note that we don't currently support non-nullable schema in file format sources because they are turned to be nullable in SQL batch code path.
Non-nullable is able to be set in SS tho. Both codes paths should be matched - it's a long standing issue.

Reply:
`catalyst` schema is always nullable when we read from the file. This is a special support for `.option("avroSchema", ...)` for Avro.

Reply:
Have we documented `avroSchema` about this the behavior?

Reply:
In my opinion, we should fix and support non-nullable in file format sources, and this can help to speed up some queries.

@viirya we can have a log as a warning.

Reply:
@dongjoon-hyun, 

> catalyst schema is always nullable when we read from the file.

This is the current status. I don't think this is expected behaviour.